File: web/src/types/ml.ts
import { v4 as uuidv4 } from 'uuid';

export enum HardwareType {
  Cpu = 'Cpu',
  Cuda = 'Cuda',
  OpenVino = 'OpenVino',
  Fallback = 'Fallback',
}

export enum WeatherType {
  Clear = 'Clear',
  Rain = 'Rain',
  Fog = 'Fog',
  Snow = 'Snow',
  Night = 'Night',
}

export interface MlEvent {
  id: string;
  event_id: string;
  truck_id: string;
  model_name: string;
  model_version: string;
  timestamp: string;
  result: MlResult;
  confidence: number;
  calibrated_confidence: number;
  latency_ms: number;
  hardware_used: HardwareType;
  meta :MlEventMetadata;
  created_at: string;
}

export type MlResult = 
  | { type: 'Drowsiness', is_drowsy: boolean, eye_closure_ratio: number }
  | { type: 'LaneDeparture', is_departing: boolean, deviation_pixels: number }
  | { type: 'CargoTamper', is_tampered: boolean, motion_score: number }
  | { type: 'LicensePlate', plate_text: string, bounding_box: [number, number, number, number] }
  | { type: 'Weather', weather_type: WeatherType, visibility_m: number }
  | { type: 'Unknown' };

export interface MlEventMetadata {
  device_id: string;
  truck_id: string;
  route_id: string;
  driver_id: string;
  camera_id: string;
  frame_timestamp: string;
  sensor_context: SensorContext | null;
  cpu_usage_percent: number;
  gpu_usage_percent: number;
  memory_used_bytes: number;
  temperature_c: number;
  model_checksum: string;
  retry_count: number;
  fallback_reason: string | null;
}

export interface SensorContext {
  speed_kmh: number;
  acceleration: number;
  steering_angle: number;
  gps_lat: number;
  gps_lon: number;
  time_of_day: string;
}

export interface MlEventSummary {
  id: string;
  event_id: string;
  truck_id: string;
  model_name: string;
  result_type: string;
  confidence: number;
  timestamp: string;
  is_alert: boolean;
  truck_license_plate: string;
  truck_model: string;
  truck_make: string;
}

export interface MlStats {
  total_events: number;
  alert_events: number;
  by_model: Record<string, number>;
  by_result: Record<string, number>;
  by_truck: Record<string, number>;
  avg_confidence: number;
  avg_latency_ms: number;
  last_24_hours: number;
  last_7_days: number;
  last_30_days: number;
}

export interface DashboardMlEvents {
  recent_ml_events: MlEventSummary[];
  stats: MlStats;
  top_models: [string, number][];
  events_by_result: Record<string, number>;
  top_trucks: [string, number][];
  time_range_hours: number;
}

File: truck-central-server/src/models/ml.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use chrono::{DateTime, Utc};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MlEvent {
    pub id: Uuid,
    pub event_id: String,
    pub truck_id: Uuid,
    pub model_name: String,
    pub model_version: String,
    pub timestamp: DateTime<Utc>,
    pub result: MlResult,
    pub confidence: f32,
    pub calibrated_confidence: f32,
    pub latency_ms: f32,
    pub hardware_used: HardwareType,
    pub meta MlEventMetadata,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum HardwareType {
    Cpu,
    Cuda,
    OpenVino,
    Fallback,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MlResult {
    Drowsiness { is_drowsy: bool, eye_closure_ratio: f32 },
    LaneDeparture { is_departing: bool, deviation_pixels: i32 },
    CargoTamper { is_tampered: bool, motion_score: f32 },
    LicensePlate { plate_text: String, bounding_box: (f32, f32, f32, f32) },
    Weather { weather_type: WeatherType, visibility_m: f32 },
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum WeatherType {
    Clear,
    Rain,
    Fog,
    Snow,
    Night,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MlEventMetadata {
    pub device_id: String,
    pub truck_id: String,
    pub route_id: String,
    pub driver_id: String,
    pub camera_id: String,
    pub frame_timestamp: DateTime<Utc>,
    pub sensor_context: Option<SensorContext>,
    pub cpu_usage_percent: f32,
    pub gpu_usage_percent: f32,
    pub memory_used_bytes: u64,
    pub temperature_c: f32,
    pub model_checksum: String,
    pub retry_count: u32,
    pub fallback_reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SensorContext {
    pub speed_kmh: f32,
    pub acceleration: f32,
    pub steering_angle: f32,
    pub gps_lat: f64,
    pub gps_lon: f64,
    pub time_of_day: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MlEventSummary {
    pub id: Uuid,
    pub event_id: String,
    pub truck_id: Uuid,
    pub model_name: String,
    pub result_type: String,
    pub confidence: f32,
    pub timestamp: DateTime<Utc>,
    pub is_alert: bool,
    pub truck_license_plate: String,
    pub truck_model: String,
    pub truck_make: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MlStats {
    pub total_events: i64,
    pub alert_events: i64,
    pub by_model: std::collections::HashMap<String, i64>,
    pub by_result: std::collections::HashMap<String, i64>,
    pub by_truck: std::collections::HashMap<Uuid, i64>,
    pub avg_confidence: f32,
    pub avg_latency_ms: f32,
    pub last_24_hours: i64,
    pub last_7_days: i64,
    pub last_30_days: i64,
}